### PR TITLE
refactor: Make useSearch hook generic and add more parameters

### PR DESCRIPTION
**Pull request by akevinge**
Addresses: #177

### CODE REVIEW

**Comment by innub** [comment]
I also think that the error from Promise should be bubbled up as the UI logic should be able to inform the user if something is wrong on our end, but I can revise it out of the PR if this is something that needs more consideration. 

I am not sure why ``AddCourseDialog`` and ``NewPlanDialog`` don't seem to be used anywhere. I'm not sure how to test if the refactor broke them, but I think they should be okay.

File: components/search/search.ts [hunk 1]
```diff
@@ -1,36 +1,51 @@
...
 /**
  * This function performs search & returns relevant data
  * */
-export default function useSearch({ getData, chips = [], filterBy = 'none' }) {
-  const [results, setResults] = React.useState([]);
+const useSearch = <T, K>({
+  getData,
+  initialQuery,
+  filterFn,
+  constraints = [0, 20],
```
Comment (ZzRanger) [inline components/search/search.ts:25]: for the constraints parameter, do you think there would be a case where the first index is not 0?

File: components/search/search.ts [hunk 2]
```diff
@@ -1,36 +1,51 @@
...
+interface SearchParams<T, K> {
+  getData: () => Promise<T[]>;
+  chips?: any[];
+  initialQuery: K;
+  filterFn: (element: T, query: K) => boolean;
+  constraints?: [number, number];
+}
+
+interface SearchReturn<T, K> {
+  results: T[];
```
Comment (ZzRanger) [inline components/search/search.ts:12]: What are your thoughts on having both results and getResults? I know it was there in the original function but do you see any use for having both?

File: components/search/search.ts [hunk 3]
```diff
@@ -1,36 +1,51 @@
 import React from 'react';
 
+interface SearchParams<T, K> {
+  getData: () => Promise<T[]>;
+  chips?: any[];
```
Comment (ZzRanger) [inline components/search/search.ts:5]: Per 9/20 discussion lets remove chip filtering for now.

**Review by ZzRanger — COMMENTED** [review]
Looks pretty good! Just left a few comments but overall great work!

To address your comments:
1. We'll handle bubbling up error when fully integrating useSearch to our existing components. I will make a separate issue for that and assign to Brady.
2. AddCourseDialog is a component for the Degree Requirement Sidebar, so if we have degree requirement validation complete we can look to integrate that component. NewPlanDialog will be modified by @Shaurya0108 .

File: components/search/search.ts (hunk 2 above)
Comment (innub) [reply components/search/search.ts:12]: I don't see any reason to have both. I wasn't sure if there was a specific reason that I wasn't aware of. Should I remove getResults?

File: components/search/search.ts (hunk 1 above)
Comment (innub) [reply components/search/search.ts:25]: Yes, 0-20 is just the default. It's up to caller if they want to choose a non-zero start index

Comment (innub) [reply components/search/search.ts:25]: Oh sorry, I see what you mean now! If we ever use ``useSearch`` for pagination maybe? Now that I think about, maybe we won't use it, but I think it's best we keep it.

File: components/search/search.ts (hunk 2 above)
Comment (ZzRanger) [reply components/search/search.ts:12]: Yup lets do that.

Comment (innub) [reply components/search/search.ts:12]: Actually, the ``usePlannerContainer`` uses getResults in a function that is no longer used. https://github.com/UTDNebula/planner/blob/543df0de5e5d400ce73bd9b55ce488db2f5ccda8/modules/planner/hooks/usePlannerContainer.ts#L17-L20

https://github.com/UTDNebula/planner/blob/543df0de5e5d400ce73bd9b55ce488db2f5ccda8/modules/planner/hooks/usePlannerContainer.ts#L87-L95 

Should I just delete the getResult parameter and the unused function in ``usePlannerContainer``?

Comment (innub) [reply components/search/search.ts:12]: Ignore the question from my last comment. We may not be able to remove getResults because the ``addItemToList``, which uses ``getResults`` is used in ``usePlannerContainer`` internally.

Comment (ZzRanger) [reply components/search/search.ts:12]: I'll take a look at this sometime this week. It shouldn't matter since the function's signature is () => T[], but I'll lyk after I do a deep dive.

Comment (innub) [reply components/search/search.ts:12]: It seems like ``getResult()`` is suppose to create a closure around  the ``result`` state. From my understanding, if we only passed in ``result:T[]``, ``addItemToList()`` would be defined with the initial value of ``result``. I don't think we can remove ``getResult()`` without breaking ``addItemToList()``.

Comment (ZzRanger) [reply components/search/search.ts:11]: Okay I tested this out w/ the following changes & it seems like addItemToList() still works when you pass in result. I think this is because every time you run updateQuery it updates result via the useState hook, which then rerenders the React hook, which triggers rerenders of the components that depend on result. 

Screen Shot 2022-09-21 at 11 37 29 PM
Screen Shot 2022-09-21 at 11 37 52 PM

That being said, I'm realizing this is probably causing a lot of unnecessary rerenders. I was thinking we could remove "results" and just have the closure, but we may need "results" to trigger updates in a component. 

tl;dr Lets keep it like this LOL

**Review by ZzRanger — APPROVED** [review]
LGTM!
